handler.go

```diff
@@ -0,0 +1,339 @@
+		return makeBidiStream(ctx, stream, nextBodyFn, responseHandlerFn)
+	case encoding.ClientStream:
+		return makeClientStream(ctx, stream, nextBodyFn, responseHandlerFn)
+	default:
+		return makeServerStream(ctx, stream, nextBodyFn, responseHandlerFn)
+	}
+}
+
+// makeServerStream starts server-side streaming rpc
+func makeServerStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	nextBodyFn streamRequestSupplierFn, responseHandlerFn streamResponseHandlerFn) error {
+
+	req, err := nextBodyFn()
+	// use nil body if no initial request input is empty, since request
+	// is mandatory in server streaming rpc
+	if err != nil && err != io.EOF {
+		return err
+	}
+	if err == nil {
+		// verify there is no second request
+		if _, err = nextBodyFn(); err == nil {
+			return fmt.Errorf("Request data contains more than 1 message for server-streaming RPC")
+		} else if err != io.EOF {
+			return err
+		}
+	}
+
+	if err = sendStreamMessage(ctx, stream, req); err != nil {
+		return err
+	}
+
+	for err == nil {
+		var resBody []byte
+		if resBody, err = receiveStreamMessage(ctx, stream); err != nil {
+			break
+		}
+
+		err = responseHandlerFn(resBody)
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+// makeClientStream starts client-side streaming rpc
+func makeClientStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	nextBodyFn streamRequestSupplierFn, responseHandlerFn streamResponseHandlerFn) error {
+
+	var err error
+	for err == nil {
+		var reqBody []byte
+		reqBody, err = nextBodyFn()
+		if err != nil {
+			break
+		}
+		err = sendStreamMessage(ctx, stream, reqBody)
+	}
+
+	if err == io.EOF {
+		err = closeSendStream(ctx, stream)
+	}
+	if err != nil {
+		return err
+	}
+
+	res, err := receiveStreamMessage(ctx, stream)
+	if err != nil {
+		return err
+	}
+	return responseHandlerFn(res)
+}
+
+// makeBidiStream starts bi-directional streaming rpc
+func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	nextBodyFn streamRequestSupplierFn, responseHandlerFn streamResponseHandlerFn) error {
+
+	var wg sync.WaitGroup
+	var sendErr error
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg.Add(1)
+	// start go routine to concurrently send stream messages
+	go func() {
+		defer wg.Done()
+
+		var err error
+		for err == nil {
+			var reqBody []byte
+			reqBody, err = nextBodyFn()
+			if err == io.EOF {
+				err = closeSendStream(ctx, stream)
+				break
+			}
+			if err != nil {
+				// cancel the context to unblock the routine waiting on receiving
+				// the stream message
+				cancel()
+				break
+			}
+
+			err = sendStreamMessage(ctx, stream, reqBody)
+		}
+
+		if err != nil {
+			sendErr = err
+		}
+	}()
+
+	var receiveErr error
+	for receiveErr == nil {
+		var resBody []byte
+		resBody, receiveErr = receiveStreamMessage(ctx, stream)
+		if receiveErr != nil {
+			break
+		}
+
+		receiveErr = responseHandlerFn(resBody)
+	}
+
+	cancel()
+	wg.Wait()
```

Comment:
I dont understand this cancel management - we have 3 cancels in this function, one after the send, one after the receive and one in a defer. Can you elaborate why you added this?

Reply:
Ok got it now :)